cfgagent/vlanconf.cpp

```diff
@@ -0,0 +1,475 @@
+#include <string.h>
+#include "logger.h"
+#include "producerstatetable.h"
+#include "macaddress.h"
+#include "vlanconf.h"
+#include "switchconf.h"
+#include "exec.h"
+#include "tokenize.h"
+
+using namespace std;
+using namespace swss;
+
+#define DOT1Q_BRIDGE_NAME   "Bridge"
+#define VLAN_PREFIX         "Vlan"
+#define LAG_PREFIX          "PortChannel"
+#define DEFAULT_VLAN_ID     1
+
+extern MacAddress gMacAddress;
+extern SwitchConf *gSwtichConfVlan;
+
+VlanConf::VlanConf(DBConnector *cfgDb, DBConnector *appDb, DBConnector *stateDb, vector<string> tableNames) :
+        OrchBase(cfgDb, tableNames),
+        m_cfgVlanTable(cfgDb, CFG_VLAN_TABLE_NAME, CONFIGDB_TABLE_NAME_SEPARATOR),
+        m_cfgVlanMemberTable(cfgDb, CFG_VLAN_MEMBER_TABLE_NAME, CONFIGDB_TABLE_NAME_SEPARATOR),
+        m_statePortTable(stateDb, STATE_PORT_TABLE_NAME, CONFIGDB_TABLE_NAME_SEPARATOR),
+        m_stateLagTable(stateDb, STATE_LAG_TABLE_NAME, CONFIGDB_TABLE_NAME_SEPARATOR),
+        m_stateVlanTable(stateDb, STATE_VLAN_TABLE_NAME, CONFIGDB_TABLE_NAME_SEPARATOR),
+        m_appVlanTableProducer(appDb, APP_VLAN_TABLE_NAME),
+        m_appVlanMemberTableProducer(appDb, APP_VLAN_MEMBER_TABLE_NAME)
+{
+    SWSS_LOG_ENTER();
+
+    // Initialize Linux dot1q bridge and enable vlan filtering
+    string cmd, res;
+
+    cmd = "ip link del ";
+    cmd += DOT1Q_BRIDGE_NAME;
+    swss::exec(cmd, res);
+    cmd = "ip link add ";
+    cmd += DOT1Q_BRIDGE_NAME;
+    cmd += " up type bridge";
+    swss::exec(cmd, res);
+    cmd = "echo 1 > /sys/class/net/";
+    cmd += DOT1Q_BRIDGE_NAME;
+    cmd += "/bridge/vlan_filtering";
+    swss::exec(cmd, res);
+    cmd = "bridge vlan del vid " + std::to_string(DEFAULT_VLAN_ID)
+            + " dev " + DOT1Q_BRIDGE_NAME + " self";
+    swss::exec(cmd, res);
+}
+
+void VlanConf::syncCfgDB()
+{
+    OrchBase::syncDB(CFG_VLAN_TABLE_NAME, m_cfgVlanTable);
+    OrchBase::syncDB(CFG_VLAN_MEMBER_TABLE_NAME, m_cfgVlanMemberTable);
+}
+
+bool VlanConf::addHostVlan(int vlan_id)
+{
+    string cmd, res;
+
+    cmd = "bridge vlan add vid " + to_string(vlan_id) + " dev "
+            + DOT1Q_BRIDGE_NAME + " self";
+    swss::exec(cmd, res);
+    cmd = "ip link add link ";
+    cmd += DOT1Q_BRIDGE_NAME;
+    cmd += " name ";
+    cmd += VLAN_PREFIX + to_string(vlan_id)
+            + " type vlan id " + to_string(vlan_id);
```

Comment:
can we contruct the command in one line, easier to ready.

Reply:
I had problem compiling it if put them in one line. Will check again to see how to around the problem.

Reply:
I think "stringstream" would be a better option here and it takes different types of input so you wudn't have to convert integers to_string

Reply:
Using stringstream looks like a good idea.